### DISCUSSION TITLE
Upgrade error for Powerlevel10k

### DISCUSSION THREAD

**Discussion by gbaker555** [Homebrew/discussions#2866] [Everyday usage] [answered]
About a week ago an upgrade to Powerlevel10k showed up on 'brew update'. However, running 'brew upgrade' produced the following error:

------------------------------------------------------------------------------------------------

==> Upgrading romkatv/powerlevel10k/powerlevel10k
  1.15.0 -> 1.16.0

Error: Your Command Line Tools are too outdated.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 13.1.

---------------------------------------------------------------------------------------------------

When this first happened, I was running Xcode 13.2 on Macos 12.1 (notice the error message asks for the Command Line Tools for Xcode 13.1). I am now on Macos 12.2 and Xcode 13.2.1 - none of the updates solved the problem. Also, this issue is specific to Powerlevel10k. When other packages show up in 'brew update' I can get them to upgrade with 'brew upgrade <packagename>'. However running 'brew upgrade' attempts to upgrade Powerlevel10k, which crashes and exits the upgrade script. 

Also, I already contacted the dev (@romkatv), who recommended I contact the Homebrew team.

Any suggestions? Workarounds? Is there a better place to post this issue? Any help appreciated.

**Comment by romkatv** [comment]
Perhaps the dependency on `zsh` causes this? I've removed this dependency (it's not really needed). Can you try again and see if it helped? You'll need to figure out how to induce `brew` to download the latest formula. I don't know how to do that.

**Comment by gbaker555** [comment]
There’s probably a more elegant way to do this, but I simply went into the '/Library/Caches/Homebrew/downloads/' folder and manually moved the already-downloaded package out of the folder, then ran ‘upgrade'. That seemed to trigger a new download:

> brew upgrade
==> Upgrading 1 outdated package:
romkatv/powerlevel10k/powerlevel10k 1.15.0 -> 1.16.0
==> Downloading https://github.com/romkatv/powerlevel10k/archive/v1.16.0.tar.gz
==> Downloading from https://codeload.github.com/romkatv/powerlevel10k/tar.gz/v1.16.0
######################################################################## 100.0% 

However, it then failed with the same error message as before.

Interesting problem.

**Comment by romkatv** [comment]
Wow. I have no idea why this package doesn't upgrade with a cryptic error message while other packages upgrade just fine. The formula is here: https://github.com/romkatv/homebrew-powerlevel10k/blob/master/Formula/powerlevel10k.rb. I'd appreciate if someone knowledgeable could help with this.

**Comment by gromgit** [comment, marked as answer]
@gbaker555, what's the full output of `brew config` and `brew doctor`?

**Reply by gbaker555** [reply]
> brew config
HOMEBREW_VERSION: 3.3.12
ORIGIN: https://github.com/Homebrew/brew
HEAD: 768b172393339b4654107dc6a840a2cd08dc64b9
Last commit: 6 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 3bbb3af1b97f4bf25aa3e9c1f360bfb2496357dd
Core tap last commit: 6 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_CORE_GIT_REMOTE: https://github.com/Homebrew/homebrew-core
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.5nltQxf8R8/org.macosforge.xquartz:0
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 13.0.0 build 1300
Git: 2.32.0 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 7.77.0 => /usr/bin/curl
macOS: 12.2-x86_64
CLT: 12.5.0.22.9
Xcode: 13.2.1

**Reply by gbaker555** [reply]
> brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Command Line Tools are too outdated.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 13.1.


Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Preferences or run:
  softwareupdate --all --install --force

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 13.1.


Warning: You have the following deprecated, cask taps tapped:
  caskroom/cask
Untap them with `brew untap`.

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ilmbase

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libtcl8.6.dylib
  /usr/local/lib/libtk8.6.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/fakemysql.h
  /usr/local/include/fakepq.h
  /usr/local/include/fakesql.h
  /usr/local/include/itcl.h
  /usr/local/include/itcl2TclOO.h
  /usr/local/include/itclDecls.h
  /usr/local/include/itclInt.h
  /usr/local/include/itclIntDecls.h
  /usr/local/include/itclMigrate2TclCore.h
  /usr/local/include/itclTclIntStubsFcn.h
  /usr/local/include/mysqlStubs.h
  /usr/local/include/odbcStubs.h
  /usr/local/include/pqStubs.h
  /usr/local/include/tcl.h
  /usr/local/include/tclDecls.h
  /usr/local/include/tclOO.h
  /usr/local/include/tclOODecls.h
  /usr/local/include/tclPlatDecls.h
  /usr/local/include/tclThread.h
  /usr/local/include/tclTomMath.h
  /usr/local/include/tclTomMathDecls.h
  /usr/local/include/tdbc.h
  /usr/local/include/tdbcDecls.h
  /usr/local/include/tdbcInt.h
  /usr/local/include/tk.h
  /usr/local/include/tkDecls.h
  /usr/local/include/tkPlatDecls.h

Warning: Unbrewed '.pc' files were found in /usr/local/lib/pkgconfig.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected '.pc' files:
  /usr/local/lib/pkgconfig/tcl.pc
  /usr/local/lib/pkgconfig/tk.pc

Warning: Unbrewed static libraries were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected static libraries:
  /usr/local/lib/libtclstub8.6.a
  /usr/local/lib/libtkstub8.6.a

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  ghostscript
  powerlevel10k

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  echo 'export PATH="/usr/local/sbin:$PATH"' >> ~/.zshrc

**Reply by gromgit** [reply]
>```
> CLT: 12.5.0.22.9
> Xcode: 13.2.1
>```

These are not the same thing:
* `Xcode` refers to the Xcode _IDE_ that you'd get from the Mac App Store, and are only really needed for a few formulae.
* `CLT` refers to the Xcode _Command Line Tools_ that Homebrew needs for more than a few formulae, and are a separate install.

The `brew doctor` message that you also saw during `brew install` is telling you to update the latter (which remains out-of-date), not the former (which is up-to-date). Do _exactly_ what it tells you, and the message should go away.

**Reply by gbaker555** [reply]
OK, this was weird. 

On the one hand, IT DID WORK: Powerlevel10k is now updated. Many thanks for the help. 

On the other hand, after being manually updated to the latest version (nuked the contents of /Library/Developer/CommandLineTools and ran sudo xcode-select --install), it still gave me the same warning. The only difference is that this time the warning was not accompanied by the exiting of the update script. 

It's been a learning experience for me as I thought the CLT were automatically updated along with Xcode. 

Thanks again for the help.

**Reply by carlocab** [reply]
> it still gave me the same warning

Did you install the CLT for Xcode 13.1? If you installed something newer, it's possible it doesn't contain the right headers for your system (yet -- Apple does that sometimes. It's weird; I know.)

But, yes, the CLT is updated separately from Xcode. If you don't actually use the Xcode.app yourself, then you'll get on just fine with having only the CLT installed.

**Reply by gbaker555** [reply]
I followed @gromgit's advice to do exactly what the error message told me - so I removed the contents of/Library/Developer/CommandLineTools, then ran 'sudo xcode-select --install'. 

Presumably, that would install the most current version of the CLT available, but I'm not so sure. 'brew config' now shows my CLT version as:

CLT: 13.0.0.0.1.1627064638

Which only makes sense if CLT doesn't share the version number of Xcode - despite the fact that the 'brew doctor' error message says: "You should download the Command Line Tools for Xcode 13.1." - as if they should be connected. 

I'm happy to dump the brew conf and doc outputs if anyone's curious. I just didn't do it before 'cause I didn't want to litter the topic with unnecessary info. I figure that if something is wrong, it's probably more a matter of macOS troubleshooting than a Homebrew problem. Let me know. 

Thanks again for the help.

**Reply by carlocab** [reply]
Yes, I believe that you might still have an outdated install. What's the output of
```
/Library/Developer/CommandLineTools/usr/bin/clang --version
pkgutil --pkg-info com.apple.pkg.CLTools_Executables
```
?

`xcode-select --install` has been known to be a little bit unreliable, so if the above commands reveal an outdated install and you want to fix the warnings, you can again remove `/Library/Developer/CommandLineTools` and download the right version from https://developer.apple.com/download/all/ instead.

If it's not giving you any issues other than the occasional warning, however, you're free to ignore it.

**Reply by gbaker555** [reply]
The versions looked outdated, so I directly downloaded the CLT for Xcode 13.2 (latest non-beta version), removed the current install, and installed from the .dmg. 

Now 'brew config' returns 

CLT: 13.2.0.0.1.1637719136

The clang --version output _has not changed_ (same file version), so it seems either it has not changed or there's something else going on. But at least now my CLTs seem to be fully up to date. Thanks for letting me know about the... um... undocumented variability of xcode-select. Good to know. 

Thanks again.

**Reply by gromgit** [reply]
> The clang --version output _has not changed_ (same file version), so it seems either it has not changed or there's something else going on.

CLT includes everything under `/Library/Developer/CommandLineTools`, so even if Clang itself isn't upgraded, the macOS SDKs, frameworks and other stuff likely has.

**Comment by romkatv** [comment]
@carlocab Do you perchance know why updating powerlevel10k was failing for the OP while other packages updated successfully? Why did updating powerlevel10k require the new xcode while updating other packages didn't?

**Reply by carlocab** [reply]
Yes, it comes from here:

https://github.com/Homebrew/brew/blob/c40b2baf9eea1776127f7d487b9b9776aafd5c05/Library/Homebrew/extend/os/mac/diagnostic.rb#L71-L77

The powerlevel10k formula does not define a `bottle` block, so `brew` thinks it requires building from source. When that happens, it runs a few checks to make sure your system is able to provide a suitable build environment.

Of course, your formula doesn't actually require compilation, but `brew` isn't quite smart enough yet to figure that out. To avoid this, you can add workflows to your tap that create `bottle` blocks when you update the formula. See the tap generated by `brew tap-new` for examples.

The build environment check isn't triggered when installing from bottles, which is the most likely explanation for why this error wasn't being triggered by other updates.
